sled-agent/src/opte/illumos/port_manager.rs

```diff
@@ -447,6 +476,131 @@ impl PortManager {
+            let rules = opte_firewall_rules(rules, port.vni(), port.mac());
+            let port_name = port_name.clone();
+            info!(
+                self.inner.log,
+                "Setting OPTE firewall rules";
+                "port" => ?&port_name,
+                "rules" => ?&rules,
+            );
+            hdl.set_fw_rules(&SetFwRulesReq { port_name, rules })?;
+        }
+        Ok(())
+    }
+}
+
+/// Translate from a slice of VPC firewall rules to a vector of OPTE rules
+/// that match the given port's MAC address. OPTE rules can only encode a
+/// single host address and protocol, so we must unroll rules with multiple
+/// hosts/protocols.
+fn opte_firewall_rules(
+    rules: &[VpcFirewallRule],
+    vni: &Vni,
+    mac: &MacAddr6,
+) -> Vec<FirewallRule> {
+    rules
+        .iter()
+        .filter(|rule| match rule.status {
+            VpcFirewallRuleStatus::Disabled => false,
+            VpcFirewallRuleStatus::Enabled => true,
+        })
+        .filter(|rule| {
+            rule.targets.is_empty() // no targets means apply everywhere
+                || rule.targets.iter().any(|nic| {
+                    // (VNI, MAC) is a unique identifier for the NIC.
+                    u32::from(nic.vni) == u32::from(*vni) && nic.mac.0 == *mac
+                })
+        })
+        .map(|rule| {
+            let priority = rule.priority.0;
+            let action = match rule.action {
+                VpcFirewallRuleAction::Allow => Action::Allow,
+                VpcFirewallRuleAction::Deny => Action::Deny,
+            };
+            let direction = match rule.direction {
+                VpcFirewallRuleDirection::Inbound => Direction::In,
+                VpcFirewallRuleDirection::Outbound => Direction::Out,
+            };
+            let ports = match rule.filter_ports {
+                Some(ref ports) if ports.len() > 0 => Ports::PortList(
+                    ports
+                        .iter()
+                        .flat_map(|range| {
+                            (range.first.0.get()..=range.last.0.get())
+                                .collect::<Vec<u16>>()
+                        })
+                        .collect::<Vec<u16>>(),
+                ),
+                _ => Ports::Any,
+            };
+            let proto_filters = rule.filter_protocols.as_ref().map_or_else(
+                || vec![ProtoFilter::Any],
+                |protos| {
+                    protos
+                        .iter()
+                        .map(|proto| {
+                            ProtoFilter::Proto(match proto {
+                                VpcFirewallRuleProtocol::Tcp => Protocol::TCP,
+                                VpcFirewallRuleProtocol::Udp => Protocol::UDP,
+                                VpcFirewallRuleProtocol::Icmp => Protocol::ICMP,
+                            })
+                        })
+                        .collect::<Vec<ProtoFilter>>()
+                },
+            );
```

Comment:
These seem to be a sequence of pretty straightforward type conversions, e.g., converting a list of ports in the control plane representation into that for OPTE. I think it might help readability (and allow unit testing) if we factor these into `From` implementations. Then this becomes a list of expressions like `let action = VpcFirewallRuleAction::from(rule.action);` etc.

Reply:
This really applies to the whole closure, actually.

Reply:
So.... I like this idea very much, and had originally intended to use it; these conversions are by far my least favorite part of this PR. The problem is Rust's "orphan rule": `only traits defined in the current crate can be implemented for types defined outside of the crate`. So the question becomes: where should such `From` implementations live? We pretty clearly don't want `oxide-vpc` to depend on `omicron-common`, and it's not obvious to me that we'd want the converse dependency, either. (But maybe it's fine?)

Another option would be to use strings as an intermediate representation, as most of the `oxide-vpc` types have a `FromStr` implementation, and most of the `omicron-common` types have a `Display` implementation. But I think in general it's a bad idea to do that, and would much rather use manual conversions.

Commit 9869832a attempts to take a middle ground: it refactors the `opte_firewall_rules` routine so that it uses a new (local) `FromVpcFirewallRule` trait that includes methods for all of the component conversions. It also moves it into a new `firewall_rules` module so that it's not cluttering up `port_manager`. LMK if you think this is better/worse/good enough. I think it's cleaner than it was, in particular since the basic logic of `opte_firewall_rules` is less obscured by the component conversions, but I'm open to further refactoring if you think it's warranted.

Reply:
I think what you have there is a solid solution, yeah. If `From` isn't possible due to orphan rules (ugh, I forgot about that), I usually go for a standalone conversion function. Since these are all POD types with no private data, that's pretty straightforward. The trait solution you have is great as well, nice idea!